pkg/operator/ceph/csi/cluster_config.go

```diff
@@ -0,0 +1,185 @@
+/*
+Copyright 2019 The Rook Authors. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package csi
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"sync"
+
+	"github.com/coreos/pkg/capnslog"
+	cephconfig "github.com/rook/rook/pkg/daemon/ceph/config"
+	"github.com/rook/rook/pkg/operator/k8sutil"
+	v1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+var (
+	logger = capnslog.NewPackageLogger("github.com/rook/rook", "ceph-csi")
+)
+
+type csiClusterConfigEntry struct {
+	ClusterID string   `json:"clusterID"`
+	Monitors  []string `json:"monitors"`
+}
+
+type csiClusterConfig []csiClusterConfigEntry
+
+// FormatCsiClusterConfig returns a json-formatted string containing
+// the cluster-to-mon mapping required to configure ceph csi.
+func FormatCsiClusterConfig(
+	clusterKey string, mons map[string]*cephconfig.MonInfo) (string, error) {
+
+	cc := make(csiClusterConfig, 1)
+	cc[0].ClusterID = clusterKey
+	cc[0].Monitors = []string{}
+	for _, m := range mons {
+		cc[0].Monitors = append(cc[0].Monitors, m.Endpoint)
+	}
+
+	ccJson, err := json.Marshal(cc)
+	if err != nil {
+		return "", fmt.Errorf("failed to marshal csi cluster config. %+v", err)
+	}
+	return string(ccJson), nil
+}
+
+func parseCsiClusterConfig(c string) (csiClusterConfig, error) {
+	var cc csiClusterConfig
+	err := json.Unmarshal([]byte(c), &cc)
+	if err != nil {
+		return cc, fmt.Errorf("failed to parse csi cluster config. %+v", err)
+	}
+	return cc, nil
+}
+
+func formatCsiClusterConfig(cc csiClusterConfig) (string, error) {
+	ccJson, err := json.Marshal(cc)
+	if err != nil {
+		return "", fmt.Errorf("failed to marshal csi cluster config. %+v", err)
+	}
+	return string(ccJson), nil
+}
+
+func monEndpoints(mons map[string]*cephconfig.MonInfo) []string {
+	endpoints := make([]string, 0)
+	for _, m := range mons {
+		endpoints = append(endpoints, m.Endpoint)
+	}
+	return endpoints
+}
+
+// UpdateCsiClusterConfig returns a json-formatted string containing
+// the cluster-to-mon mapping required to configure ceph csi.
+func UpdateCsiClusterConfig(
+	curr, clusterKey string, mons map[string]*cephconfig.MonInfo) (string, error) {
+
+	var (
+		cc     csiClusterConfig
+		centry csiClusterConfigEntry
+		found  bool
+	)
+	cc, err := parseCsiClusterConfig(curr)
+	if err != nil {
+		return "", err
```

Comment:
please name the error, format it with an understandable message.

Reply:
I'm sorry but I don't understand what you mean by this.

Reply:
Sorry I meant instead of returning `err`, could you please do something like, `fmt.Errorf("my error message. %+v", err)`